### PR TITLE
Load default loadflow parameters from json file in classpath 

### PR DESCRIPTION
**Pull request by HugoKulesza**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?**

A new way to define default loadflow parameters is added.

The order of loadflow parameters override is now the following :
- Default parameters as written in the LoadFlowParameters class
- (new) Optional parameters defined in a JSON file in the classpath, linked using an implementation of LoadFlowDefaultParametersLoader
- Optional parameters defined in the powsybl config.yml file of the user 

Only one parameters JSON file can be used, if multiple are found none are used to override parameters

### CODE REVIEW

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/AbstractLoadFlowDefaultParametersLoader.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * @author Hugo Kulesza {@literal <hugo.kulesza at rte-france.com>}
+ */
+public abstract class AbstractLoadFlowDefaultParametersLoader implements LoadFlowDefaultParametersLoader {
+
+    private final String name;
+    private final String jsonParametersFile;
+
+    AbstractLoadFlowDefaultParametersLoader(String name, String jsonParametersFile) {
+        this.name = name;
```
Comment (geofjamg) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/AbstractLoadFlowDefaultParametersLoader.java:21]: Objects.requireNonNull ?

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java [hunk 2]
```diff
@@ -182,6 +187,25 @@ protected static void load(LoadFlowParameters parameters, PlatformConfig platfor
...
 
     public LoadFlowParameters() {
+        this(ServiceLoader.load(LoadFlowDefaultParametersLoader.class)
+                .stream()
+                .map(ServiceLoader.Provider::get)
+                .collect(Collectors.toList()));
+    }
+
+    public LoadFlowParameters(List<LoadFlowDefaultParametersLoader> defaultParametersLoaders) {
+        int numberOfLoadersFound = defaultParametersLoaders.size();
```
Comment (geofjamg) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java:197]: Same

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java [hunk 3]
```diff
@@ -182,6 +187,25 @@ protected static void load(LoadFlowParameters parameters, PlatformConfig platfor
...
+        if (numberOfLoadersFound > 1) {
+            List<String> names = defaultParametersLoaders.stream()
+                    .map(LoadFlowDefaultParametersLoader::getSourceName)
+                    .collect(Collectors.toList());
+            LOGGER.warn("Multiple default loadflow parameters classes have been found in the class path : {}. No default parameters file loaded",
+                    names);
+        } else if (numberOfLoadersFound == 1) {
+            LoadFlowDefaultParametersLoader loader = defaultParametersLoaders.get(0);
+            JsonLoadFlowParameters.update(this, loader.loadDefaultParametersFromFile());
+            LOGGER.info("Default loadflow configuration has been updated using the reference file from parameters loader '{}'", loader.getSourceName());
```
Comment (geofjamg) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java:207]: Maybe just a debug log

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java [hunk 4]
```diff
@@ -182,6 +187,25 @@ protected static void load(LoadFlowParameters parameters, PlatformConfig platfor
     private double dcPowerFactor = DEFAULT_DC_POWER_FACTOR;
 
     public LoadFlowParameters() {
+        this(ServiceLoader.load(LoadFlowDefaultParametersLoader.class)
+                .stream()
+                .map(ServiceLoader.Provider::get)
+                .collect(Collectors.toList()));
```
Comment (geofjamg) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java:193]: toList()

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java [hunk 5]
```diff
@@ -182,6 +187,25 @@ protected static void load(LoadFlowParameters parameters, PlatformConfig platfor
...
+    }
+
+    public LoadFlowParameters(List<LoadFlowDefaultParametersLoader> defaultParametersLoaders) {
+        int numberOfLoadersFound = defaultParametersLoaders.size();
+        if (numberOfLoadersFound > 1) {
+            List<String> names = defaultParametersLoaders.stream()
+                    .map(LoadFlowDefaultParametersLoader::getSourceName)
+                    .collect(Collectors.toList());
```
Comment (geofjamg) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlowParameters.java:201]: toList()

File: loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/AbstractLoadFlowDefaultParametersLoader.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        this.jsonParametersFile = jsonParametersFile;
```
Comment (geofjamg) [inline loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/AbstractLoadFlowDefaultParametersLoader.java:22]: same

**Review by geofjamg — COMMENTED** [review]
Some minor issues

**Review by geofjamg — APPROVED** [review]

File: loadflow/loadflow-api/src/test/java/com/powsybl/loadflow/LoadFlowDefaultParametersLoaderMock.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+package com.powsybl.loadflow;
+
+/**
+ * @author Hugo Kulesza {@literal <hugo.kulesza at rte-france.com>}
+ */
+public class LoadFlowDefaultParametersLoaderMock extends AbstractLoadFlowDefaultParametersLoader {
```
Comment (flo-dup) [inline loadflow/loadflow-api/src/test/java/com/powsybl/loadflow/LoadFlowDefaultParametersLoaderMock.java:13]: Couldn't you directly test it by adding `@AutoService(LoadFlowDefaultParametersLoader.class)`?

Comment (flo-dup) [reply loadflow/loadflow-api/src/test/java/com/powsybl/loadflow/LoadFlowDefaultParametersLoaderMock.java:13]: My bad, you need two classes with empty constructor then.

Comment (flo-dup) [reply loadflow/loadflow-api/src/test/java/com/powsybl/loadflow/LoadFlowDefaultParametersLoaderMock.java:13]: and it is bad practice in fact!

**Review by flo-dup — APPROVED** [review]

**Pull request merged**